src/lib/projectConfigValidation.ts

```diff
@@ -0,0 +1,437 @@
+  }
+}
+
+async function validateDocumentDiagnostics(
+  document: TextDocument
+): Promise<Diagnostic[]> {
+  if (!document.fileName.endsWith('-hsmeta.json')) return [];
+  const projectRoot = await findProjectRoot(document.uri);
+  if (!projectRoot) return [];
+  const projectVersion = await getProjectVersion(projectRoot);
+  if (!projectVersion) return [];
+  // Any file that fails the above checks is one that we don't handle
+  let metaFile: any; // The type is whatever JSON the user provided us with
+  try {
+    metaFile = JSON.parse(document.getText());
+  } catch (error) {
+    return [
+      new Diagnostic(
+        new Range(0, 0, 0, 0),
+        errorMessages.validation.invalidJson,
+        DiagnosticSeverity.Error
+      ),
+    ];
+  }
+  const root = jsonc.parseTree(document.getText());
+  if (!root) {
+    // Should not happen, we've already parsed it above without jsonc
+    return [
+      new Diagnostic(
+        new Range(0, 0, 0, 0),
+        errorMessages.validation.invalidJson,
+        DiagnosticSeverity.Error
+      ),
+    ];
+  }
+  if (!('type' in metaFile)) {
+    return [
+      new Diagnostic(
+        new Range(0, 0, 0, 0),
+        errorMessages.validation.missingType,
+        DiagnosticSeverity.Error
+      ),
+    ];
+  }
+  const uidError = validateUid(metaFile.uid);
+  if (uidError) {
+    const uidNode = jsonc.findNodeAtLocation(root, ['uid']);
+    return [
+      new Diagnostic(
+        uidNode
+          ? new Range(
+              document.positionAt(uidNode.offset),
+              document.positionAt(uidNode.offset + uidNode.length)
+            )
+          : new Range(0, 0, 0, 0),
+        uidError,
+        DiagnosticSeverity.Error
+      ),
+    ];
+  }
+  if (!('config' in metaFile)) {
+    return [
+      new Diagnostic(
+        new Range(0, 0, 0, 0),
+        errorMessages.validation.missingConfig,
+        DiagnosticSeverity.Error
+      ),
+    ];
+  }
+  const accountId = getAccountId();
+  if (!accountId) {
+    return [
+      new Diagnostic(
+        new Range(0, 0, 0, 0),
+        `${errorMessages.api.accountIdIsRequiredToFetchSchemas}. Authenticate account with the HubSpot CLI`,
+        DiagnosticSeverity.Error
+      ),
+    ];
+  }
+  const validator = await getCachedValidatorForExternalType(
+    projectRoot,
+    projectVersion,
+    accountId,
+    metaFile.type
+  );
+  if ('error' in validator) {
+    return [
+      new Diagnostic(
+        new Range(0, 0, 0, 0),
+        validator.error,
+        DiagnosticSeverity.Error
+      ),
+    ];
+  }
+  if (!validator(metaFile['config'])) {
+    let diagnostics: Record<string, Diagnostic> = {};
+    for (const error of validator.errors || []) {
+      const { message, nodePath, onlyKeyNode } = customizeAjvError(error);
+      if (!message) {
+        continue;
+      }
+      // This is a duplicate error, skip it
+      if (
+        diagnosticDeduplicationKey(nodePath, error.keyword, error.params) in
```

Comment:
Slight nit, but we call `diagnosticDeduplicationKey` three times in here. Maybe we can call it just once at the start?